### ISSUE TITLE
OOM issue on Ubuntu while bulk loading Freebase

### ISSUE THREAD

**Issue by kushalarora** [openlink/virtuoso-opensource#149]
I am trying to create a local Sparql endpoint for Freebase for running some local experiments. While using virtuoso 7, I regularly see server getting killed due by OOM killer. I have followed all the required steps as mentioned here. I have also made the required changes to my virtuso.ini file as mentioned in RDF Performance Tuning.

My system configuration is:

8 CPU 2.9 Ghz 16 GB RAM.

I have enough harddisk too.

Regarding data dumps, I have splitted the freebase data dump(23GB gzipped, aprrox 250 GB uncompressed) into 10 smaller gzipped files containing 200000000 triples each.

Following are the changes I made to virtuoso.ini

NumberOfBuffers = 1360000
MaxDirtyBuffers = 1000000

MaxCheckpointRemap = 340000 (1/4th of NumberOfBuffers)

Along with this I have set vm.swapiness = 10 as mentioned in 2.

Am I missing something obvious?

P.S.: I did try virrtuoso-opensource-6.1 too. But it appeared to be too slow.

One interesting observation I had was that during bulk loading process virtuoso-6.1 memory consumption was rising too slowly, but it might be because general indexing itself was too slow.

Another observation I had was the virtuoso-6.1 at start time occupies almost negligible memory(order of 500MB's) whereas virtuoso-7 starts with approx 6500 MB and grows quickly.

**Comment by HughWilliams** [comment]
What method are you using to perform the load, are you using the Virtuoso RDF Bulk Loader functions as detailed at:

http://virtuoso.openlinksw.com/dataspace/doc/dav/wiki/Main/VirtBulkRDFLoader

**Comment by kushalarora** [comment]
yes. I am exactly using the method detailed in the link mentioned above.

**Comment by HughWilliams** [comment]
OK, what is the Virtuoso 7 version being run, please provide the output of  status('') run from isql and/or virtuoso-t -? , as would like to ensure you are using the latest git develop/7 branch ?

Also, are the Freebase split files you are loading available for download and we could try this in-house ?

**Comment by kushalarora** [comment]
I initially tried with develop7 branch, but now moved to stable 7 
branch. I can try develop7 branch again. I can host the split files.

You can find splitted, gzipped freebase dump files at 
http://149.165.158.57:3035/

Regards
Kushal Arora

**Comment by kushalarora** [comment]
Output of status is as follows:

Running Statements:
  Time (msec) Text
     22318714 rdf_loader_run(log_enable=>3)
     22318715 rdf_loader_run(log_enable=>3)
     22318716 rdf_loader_run(log_enable=>3)
     22318715 rdf_loader_run(log_enable=>3)
     22318717 rdf_loader_run(log_enable=>3)
     22318729 rdf_loader_run(log_enable=>3)
     22318721 rdf_loader_run(log_enable=>3)
     22318717 rdf_loader_run(log_enable=>3)
          234 #line 1 "(console)"
status()

Hash indexes

92 Rows. -- 235 msec.

**Comment by HughWilliams** [comment]
I cannot access http://149.165.158.57:3035/  ?

**Comment by kushalarora** [comment]
Apologies for inconvinence. Thee server died. Can you try now. It is up now.

**Comment by HughWilliams** [comment]
The files were downloading and then suddenly the connection was loss so the server seems to have gone down again:

# wget http://149.165.158.57:3035/xaa.gz

--2014-03-15 03:10:29--  http://149.165.158.57:3035/xaa.gz
Connecting to 149.165.158.57:3035... Sat Mar 15 03:10:29 2014 wget[28630] libsdp Error connect: for TCP fd:4 failed with error:Connection refused
 failed: Connection refused.

**Comment by kushalarora** [comment]
Apologies. There was some issue. Ill ensure it stays up now.

**Comment by kushalarora** [comment]
Hello @HughWilliams . Were you able to reproduce the issue?

**Comment by HughWilliams** [comment]
I wasn't able to recreate the OOM error but we did find that the Freebase datasets contain many duplicate triples and contain large literal values which consume much more space than the average size and this consumes significantly more space that the 10bytes per quad average for v7. 

The end result  being that the buffers allocated for Virtuoso are consume very quickly resulting in the data having to be swapped in and out of disk during the load which then grinds it pretty much to a halt (well a very slow load rate 100 or 1000 of triples per sec rather then 100,000+). Development are currently performing some more analysis on the dataset before concluding there findings.

Will let you know what there final analysis is ....

What is the max memory available on the machine you are using ?

**Comment by kushalarora** [comment]
I have approx 16GB of memory available.
